plugins/module_utils/client/discovery.py

```diff
@@ -0,0 +1,160 @@
+# Copyright [2017] [Red Hat, Inc.]
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+import json
+import os
+from collections import defaultdict
+import hashlib
+import tempfile
+
+import kubernetes.dynamic
+import kubernetes.dynamic.discovery
+from kubernetes import __version__
+from kubernetes.dynamic.exceptions import ServiceUnavailableError
+
+from ansible.module_utils import six
+from ansible_collections.kubernetes.core.plugins.module_utils.client.resource import ResourceList
+
+
+class Discoverer(kubernetes.dynamic.discovery.Discoverer):
+    def __init__(self, client, cache_file):
+        self.client = client
+        default_cache_file_name = 'k8srcp-{0}.json'.format(hashlib.sha256(self.__get_default_cache_id()).hexdigest())
+        self.__cache_file = cache_file or os.path.join(tempfile.gettempdir(), default_cache_file_name)
+        self.__init_cache()
+
+    def __get_default_cache_id(self):
+        user = self.__get_user()
+        if user:
+            cache_id = "{0}-{1}".format(self.client.configuration.host, user)
+        else:
+            cache_id = self.client.configuration.host
+
+        if six.PY3:
+            return cache_id.encode('utf-8')
+        return cache_id
+
+    def __get_user(self):
+        if hasattr(os, 'getlogin'):
```

Comment:
Why do we need this construction. As far as I can see, `getlogin` is always defined.

Reply:
Also, the documentation advice to use `https://docs.python.org/3.6/library/getpass.html#getpass.getuser` which actually can also do the environment variable look up too.

Reply:
This was a copy/paste from the openshift client. From what I can tell it was added to provide support for Windows. I agree it could probably be changed, but I'm hesitant to do so considering we don't do any portability testing and I have no easy way to test any change locally on Windows. Ideally, this code is going to be removed once we can get these or similar changes upstream in the kubernetes client.

Reply:
May I suggest to add a comment in the code about that.